src/middleware/compression.rs

```diff
@@ -0,0 +1,233 @@
+    }
+
+    fn preferred_encoding(&self, headers: &HeaderMap) -> Result<Encoding, Error> {
+        let mut choices: Vec<(Encoding, f32)> = Vec::new();
+        for hval in headers.get_all(ACCEPT_ENCODING).iter() {
+            for cap in RE_ACCEPT_ENCODING.captures_iter(hval.to_str().unwrap()) {
+                match self.parse_encoding(cap.name("a").unwrap().as_str()) {
+                    Ok(encoding) => match cap.name("q") {
+                        Some(qval) => {
+                            let qval = qval.as_str().parse::<f32>().unwrap();
+                            if qval == 0. && encoding == Encoding::Identity {
+                                return Err(Error::from(StatusCode::NOT_ACCEPTABLE));
+                            } else if qval > 0. {
+                                choices.push((encoding, qval))
+                            }
+                        }
+                        None => choices.push((encoding, 1.0)),
+                    },
+                    Err(e) => continue, // ignore encodings we can't perform
+                }
+            }
+        }
+
+        choices.sort_by(|(_, q1), (_, q2)| q2.partial_cmp(&q1).unwrap());
+
+        if !choices.is_empty() {
+            Ok(choices[0].0)
+        } else {
+            Ok(Encoding::Identity)
+        }
+    }
+    /// Consumes the response and returns an encoded version of it.
+    fn encode(&self, mut res: Response, encoding: Encoding) -> Response {
+        // TODO: test and make sure this works
+        res.headers_mut()
+            .append(CONTENT_ENCODING, encoding.header_value());
+        match encoding {
+            Encoding::Gzip => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let level = self.settings.gzip_level;
+                let stream = body.map(move |b| match b {
+                    Ok(b) => {
+                        let reader = b.into_buf().reader();
+                        let mut buffer: Vec<u8> = Vec::new();
+                        let mut encoder = GzEncoder::new(reader, level);
+                        match encoder.read_to_end(&mut buffer) {
+                            Ok(_) => Ok(Bytes::from(buffer)),
+                            Err(e) => Err(e),
+                        }
+                    }
+                    Err(e) => Err(e),
+                });
+                *res.body_mut() = Body::from_stream(stream);
+                res
+            }
+            Encoding::Deflate => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let level = self.settings.deflate_level;
+                let stream = body.map(move |b| match b {
+                    Ok(b) => {
+                        let reader = b.into_buf().reader();
+                        let mut buffer: Vec<u8> = Vec::new();
+                        let mut encoder = DeflateEncoder::new(reader, level);
```

Comment:
Creating a new encoder for each chunk is not going to work, they implicitly create internal state as they’re encoding the data, and later chunks will depend on this state from earlier chunks. You will need to use the [low-level in memory chunk encoder](https://docs.rs/flate2/1.0.7/flate2/struct.Compress.html) and keep one instance around for the entire stream.

Reply:
Alright, noted. I may need help with this implementation as I am still pretty wet behind the ears when it comes to Stream wrappers. Will try my best though.

Reply:
You can ping me on either discord if you've got questions.

Reply:
Will do, thank you!